### PR TITLE
Block cross-package skips/replaces on add

### PR DESCRIPTION
**Pull request by ankitathomas**
add validation to ensure bundles only skip/replace members of the same package

Signed-off-by: Ankita Thomas <ankithom@redhat.com>

### CODE REVIEW

File: pkg/registry/populator.go [hunk 1]
```diff
@@ -407,3 +411,54 @@ func DecodeFile(path string, into interface{}) error {
...
+	pkgs, err := i.querier.ListPackages(context.TODO())
+	if err != nil {
+		return fmt.Errorf("unable to verify bundle packages: %v", err)
+	}
+	for _, pkg := range pkgs {
+		entries, err := i.querier.GetChannelEntriesFromPackage(context.TODO(), pkg)
+		if err != nil {
+			return fmt.Errorf("unable to verify bundles for package %v", err)
+		}
+		fmt.Println(expectedBundlePackages)
```
Comment (dinhxuanvu) [inline pkg/registry/populator.go:455]: I assume this is test code so it should be removed.

**Review by dinhxuanvu — COMMENTED** [review]
Just a few nits. Looking good.

File: pkg/registry/populator.go (hunk 1 above)
Comment (ankitathomas) [reply pkg/registry/populator.go]: removed

File: go.mod [hunk 2]
```diff
@@ -5,6 +5,7 @@ go 1.16
 require (
 	github.com/Microsoft/hcsshim v0.8.9 // indirect
 	github.com/blang/semver v3.5.1+incompatible
```
[the remark this answers is not in the document]
Comment (ankitathomas) [reply go.mod]: Added a new commit with a general version bump throughout operator-registry.

**Review by dinhxuanvu — APPROVED** [review]
Looking good.
/approve

File: pkg/registry/populator.go [hunk 3]
```diff
@@ -407,3 +411,54 @@ func DecodeFile(path string, into interface{}) error {
...
+				return fmt.Errorf("bundle %s must belong to exactly one package, found on: %v", bndl, pkgs)
+			}
+			expectedBundlePackages[bndl] = b.Bundle.Package
+		}
+	}
+	if len(expectedBundlePackages) == 0 {
+		return nil
+	}
+
+	pkgs, err := i.querier.ListPackages(context.TODO())
```
Comment (anik120) [inline pkg/registry/populator.go:447]: Ideally, we'd place this kind of check in the table schema itself(eg put a constraint in the replaces/skips fields in the operatorbundle table to allow bundle names that belong to the same package only). Calling the querier, i.e opening a db connection every single time is an expensive operation. Opening a db connection for every additional step we take while loading a bundle = making the load operation a little bit more inefficient every time. 

Having said that, if this is being considered as a quick dirty fix for the time sqllite operations are deprecated, and we're okay with the load operation being a little bit inefficient, consider this comment withdrawn.

File: pkg/registry/populator_test.go [hunk 4]
```diff
@@ -3068,3 +3071,86 @@ func newUnpackedTestBundle(root, dir, name string, csvSpec json.RawMessage, anno
...
+		require.NoError(t, err)
+
+		b, err := registry.NewBundleFromStrings(name, versionString, pkg, defaultChannel, channels, string(jsonout))
+		require.NoError(t, err)
+		return &registry.ImageInput{Bundle: b}
+	}
+
+	logrus.SetLevel(logrus.DebugLevel)
+	db, _ := CreateTestDb(t)
+	//	defer cleanup()
```
Comment (anik120) [inline pkg/registry/populator_test.go:3112]: I'm guessing this was meant to be uncommented.

File: go.mod [hunk 5]
```diff
@@ -4,7 +4,7 @@ go 1.16
 
 require (
 	github.com/Microsoft/hcsshim v0.8.9 // indirect
-	github.com/blang/semver v3.5.1+incompatible
+	github.com/blang/semver/v4 v4.0.0
```
Comment (anik120) [inline go.mod:7]: Why was this necessary?

**Review by anik120 — CHANGES_REQUESTED** [review]

File: pkg/registry/populator.go (hunk 3 above)
Comment (ankitathomas) [reply pkg/registry/populator.go:447]: We could add it to schema, but we'd also need to enforce the one package per bundle rule or this still happens. It's not just making sure it's in the right package, but that it isn't in any other package.

File: go.mod (hunk 5 above)
Comment (ankitathomas) [reply go.mod:7]: The `Version` in `ClusterServiceVersionSpec` requires semver/v4. It didn't make sense to keep both semver versions around, and since operator-lifecycle-manager already uses v4 it seemed like a good choice to keep.

**Review by dinhxuanvu — APPROVED** [review]
/approve

File: pkg/registry/populator.go (hunk 3 above)
Comment (anik120) [reply pkg/registry/populator.go:447]: Shouldn't we be doing that anyway then?

File: go.mod (hunk 5 above)
Comment (anik120) [reply go.mod:7]: Could you link where exactly you need `semver/v4` in the `ClusterServiceVersionSpec` for this PR?

Comment (ankitathomas) [reply go.mod:7]: https://github.com/operator-framework/operator-registry/pull/826/files#diff-bc4c211037518ff105dee0994c7e5fe4ae00bfd9e3230cb17b82605894dfc33aR3083

That requires `OperatorVersion`, which is `semver/v4`

Comment (anik120) [reply go.mod:7]: Okay that makes sense.

File: pkg/registry/populator.go (hunk 3 above)
Comment (ankitathomas) [reply pkg/registry/populator.go:447]: We can have a trigger checking for this but with the way adding bundles in sqlite works, that alone would get us bundles that are partially added (i.e, in `operatorbundle` table alone with some entries into `channel_entry` table). Considering we also nuke the package for reconstructing bundle graphs, having that start and fail half way would be really bad.

So we'd need to do this check before starting the add anyway, since our bundle adds aren't one atomic operation.

Comment (kevinrizza) [reply pkg/registry/populator.go:447]: I don't think we should touch the schema or make a db migration unless we absolutely have to

Comment (anik120) [reply pkg/registry/populator.go:447]: That's a good point too.

**Review by anik120 — APPROVED** [review]
/lgtm
